src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -743,7 +780,9 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message,
                         return 0;
 
                 /* for now pick a random free address from the pool */
-                if (existing_lease)
+                if (static_lease)
+                        address = static_lease->address;
+                else if (existing_lease)
                         address = existing_lease->address;
                 else {
                         struct siphash state;
```

Comment:
When no matching static and existing lease does not exist, we need to check that the generated address in this `else` branch must not be listed in static leases.